python/desc/sims/GCRCatSimInterface/ProtoDC2DatabaseEmulator.py

```diff
@@ -291,43 +281,103 @@ def _postprocess_results(self, master_chunk):
-        gid_max = np.nanmax(gid_arr)
+
+        self._cached_half_space = half_space
+        trixel_bounds = half_space.findAllTrixels(6)
 
         query = 'SELECT galaxy_id, magNorm, varParamStr '
         query += 'FROM agn_params '
-        query += 'WHERE galaxy_id BETWEEN %d AND %d ' % (gid_min, gid_max)
+        query += 'WHERE '
+        for i_bound, bound in enumerate(trixel_bounds):
+            if i_bound>0:
+                query += 'OR '
+            if bound[0]==bound[1]:
+                query += 'htmid_6 == %d ' % bound[0]
+            else:
+                query += '(htmid_6 >= %d AND htmid_6 <= %d) ' % (bound[0], bound[1])
         query += 'ORDER BY galaxy_id'
 
-        agn_data_iter = self._agn_dbo.get_arbitrary_chunk_iterator(query,
-                                                        dtype=self._agn_dtype,
-                                                        chunk_size=1000000)
+        self._agn_query_results = self._agn_dbo.execute_arbitrary(query,
+                                                                  dtype=self._agn_dtype)
+
+
+    def _prefilter_galaxy_id(self, obs_metadata):
+        """
+        Accept an ObservationMetaData characterizing
+        the current pointing.
+
+        Return a numpy array of galaxy_ids that are in the
+        field of view and actually contain an AGN.
+        """
+
+        half_space = halfSpaceFromRaDec(obs_metadata.pointingRA,
+                                        obs_metadata.pointingDec,
+                                        obs_metadata.boundLength)
+
+        if (not hasattr(self, "_agn_query_results") or
+            half_space != self._cached_half_space):
+
+            self._do_agn_query(half_space)
+
+        return np.sort(self._agn_query_results['galaxy_id'])
 
 
+
+
+    def _postprocess_results(self, master_chunk, obs_metadata):
+        """
+        query the database specified by agn_params_db to
+        find the AGN varParamStr associated with each AGN
+        """
+
+        if self.agn_objid is None:
+            gid_name = 'galaxy_id'
+            varpar_name = 'varParamStr'
+            magnorm_name = 'magNorm'
+        else:
+            gid_name = self.agn_objid + '_' + 'galaxy_id'
+            varpar_name = self.agn_objid + '_' + 'varParamStr'
+            magnorm_name = self.agn_objid + '_' + 'magNorm'
+
+        if self.agn_params_db is None:
+            return(master_chunk)
+
+        half_space = halfSpaceFromRaDec(obs_metadata.pointingRA,
+                                        obs_metadata.pointingDec,
+                                        obs_metadata.boundLength)
+
+        if (not hasattr(self, "_agn_query_results") or
+            half_space != self._cached_half_space):
+
+            self._do_agn_query(half_space)
+
+        gid_arr = master_chunk[gid_name]
         m_sorted_dex = np.argsort(gid_arr)
         m_sorted_id = gid_arr[m_sorted_dex]
-        for agn_chunk in agn_data_iter:
```

Comment:
This iteration is no longer needed because it's taken care by `_do_agn_query`, I guess?

Reply:
True.

More specifically, it is no longer needed because `_do_agn_query` is now loading all of the AGN at once into one giant chunk (it is actually not giant; there are only 30 million AGN, as opposed to the few billion galaxies in cosmoDC2).